hilti/toolchain/include/ast/types/auto.h

```diff
@@ -0,0 +1,62 @@
+// Copyright (c) 2020 by the Zeek Project. See LICENSE for details.
+
+#pragma once
+
+#include <memory>
+#include <utility>
+#include <vector>
+
+#include <hilti/ast/type.h>
+#include <hilti/ast/types/unknown.h>
+
+namespace hilti {
+namespace type {
+
+/** AST node for an "auto" type. */
+class Auto : public TypeBase,
```

Comment:
Well, it's complicated. :) The reason for adding `auto` is purely technical: it helps getting around some cyclic type dependencies. One of the later commits uses it for the `$$` hook parameters. I wasn't able to get them to reliably resolve without this new `auto` mechanism--which is nicer anyways because it removes some duplication too.

I don't think replacing this with `type::Unknown` would work right now because generally, type resolution doesn't replace such nodes explicitly. `type::Unknown` is more of a temporary stand-in until we have means to resolve the type otherwise (e.g., some `type()` methods that are dynamically calculating a type return `Unknown` until they can do so). `Auto` on the other hand gives us a specific node to search and replace/update.

Honestly, our type resolution isn't great right now: it has diverged into a mix of multiple approaches over time, and `auto` is certainly a bit of hack here to make things work. I do like the idea of `auto` conceptuallly, but it would be better to use it consistently everywhere where we the language infers types by itself (e.g., `local x = 42` would use it too). But overall, the real task is to clean up type resolution across the AST infrastructure, and probably more along the lines of what you have in mind: always store types in explicit nodes (which could then be `type::Unknown`) and have dynamic computation replace those.

In the meantime I think this will do (and it's HILTI-only, not exposed to Spicy users).